thirdparty/hairer/radau5_superlu_complex.c

```diff
@@ -0,0 +1,234 @@
+    slu_aux->drop_tol = 0.0; // Default, not implemented for pdgstrf
+    slu_aux->lwork = 0; // flag; work-space allocated internally
+    slu_aux->work = NULL; // internal work space; not referenced due to lwork = 0
+
+    StatAlloc(slu_aux->n, slu_aux->nprocs, slu_aux->panel_size, slu_aux->relax, slu_aux->Gstat);
+    StatInit(slu_aux->n, slu_aux->nprocs, slu_aux->Gstat);
+
+    slu_aux->perm_r = intMalloc(slu_aux->n);
+    slu_aux->perm_c = intMalloc(slu_aux->n);
+    slu_aux->rhs = doublecomplexMalloc(slu_aux->n);
+
+    if (!slu_aux->perm_r) {SUPERLU_ABORT("Malloc failed for perm_r[].");}
+    if (!slu_aux->perm_c) {SUPERLU_ABORT("Malloc failed for perm_c[].");}
+    if (!slu_aux->rhs)    {SUPERLU_ABORT("Malloc failed for rhs[].");}
+
+    zCreate_Dense_Matrix(slu_aux->B, slu_aux->n, 1, slu_aux->rhs, slu_aux->n, SLU_DN, SLU_Z, SLU_GE);
+
+    // allocate memory for storing matrix of linear system
+    // min(nnz_jac + n, n*n) is upper bound on storage requirement of linear system
+    slu_aux->data_sys = doublecomplexMalloc(min(slu_aux->nnz_jac + slu_aux->n, n*n));
+    if (!slu_aux->data_sys)    {SUPERLU_ABORT("Malloc fails for data_sys[].");}
+
+    return slu_aux;
+}
+
+// Setting up the matrix to be factorized
+int superlu_setup_z(SuperLU_aux_z *slu_aux, double scale_r, double scale_i,
+                    double *data_J, int *indices_J, int *indptr_J,
+                    int fresh_jacobian, int jac_nnz){
+    NCformat *AStore = slu_aux->A->Store;
+    SUPERLU_FREE(AStore);
+
+    // number of non-zero elements maz have changed during recent jacobian evaluation
+    slu_aux -> nnz_jac = jac_nnz;
+
+    int current_idx = 0;
+    int i, j;
+
+    // build system matrix (scale_r + i*scale_i) * I - JAC
+    // Copy jacobian data to slu_aux structure & scale diagonal
+    slu_aux->indices_sys = indices_J;
+    slu_aux->indptr_sys = indptr_J;
+
+    for (i = 0; i < slu_aux-> n; i++){
+        for (j = indptr_J[i]; j < indptr_J[i+1]; j++){
+            slu_aux->data_sys[current_idx].r = -data_J[current_idx];
+            slu_aux->data_sys[current_idx].i = 0;
+            if (i == indices_J[current_idx]){
+                slu_aux->data_sys[current_idx].r = slu_aux->data_sys[current_idx].r + scale_r;
+                slu_aux->data_sys[current_idx].i = scale_i;
+            }
+            current_idx++;
+        }
+    }
+    
+    zCreate_CompCol_Matrix(slu_aux->A, slu_aux->n, slu_aux->n, slu_aux->nnz_jac,
+                           slu_aux->data_sys, slu_aux->indices_sys, slu_aux->indptr_sys,
+                           SLU_NC, SLU_Z, SLU_GE);
+
+    if (fresh_jacobian){
+        get_perm_c(3, slu_aux->A, slu_aux->perm_c); // 3 = approximate minimum degree for unsymmetrical matrices
+        slu_aux->refact = NO; // new jacobian -> number of elements may have changed, re-using factorization may no longer be valid
```

Comment:
Is the comment correct? It seems to indicate that if the structure has not changed we can keep the factorization? If so, then we should have a separate check that look at the structure and see if it has changed or not.

Reply:
After some discussion, we decided to keep this for now. This might be subject to future investigation.